src/common/pool_map.c

```diff
@@ -1373,26 +1363,91 @@ uuid_compare_cb(const void *a, const void *b)
 	return uuid_compare(*ua, *ub);
 }
 
+static int
+add_domains_to_pool_buf(struct pool_map *map, struct pool_buf *map_buf,
+			int map_version,
+			int ndomains, const uint32_t *domains)
+{
+	int		i, rc;
+	uint32_t	num_comps;
+	uint8_t		new_status;
+	const int	TUPLE_SIZE = POOL_BUF_DOMAIN_TUPLE_LEN;
+	uint32_t	domains_found;
+
+	if (map != NULL) {
+		new_status = PO_COMP_ST_NEW;
+		num_comps = pool_map_find_domain(map, PO_COMP_TP_RACK,
+						 PO_COMP_ID_ALL, NULL);
+	} else {
+		new_status = PO_COMP_ST_UPIN;
+		num_comps = 0;
+	}
+
+	/*
+	 * Fault domains are structured as an array of tuples:
+	 * (layer, ID, number of children)
+	 * The leaves are simply the rank IDs, and may be ignored for now.
+	 */
+	domains_found = 1 + domains[TUPLE_SIZE - 1];
```

Comment:
So the number of logical domain components must be >= 1 is the assumption? Since the format of domains is not so fool proof, we might consider checking and/or asserting (if the case shall happen only due to bugs) some constraints.

Reply:
Actually, would it be feasible and reasonable to change the uint32_t domains[] parameters to void *domains, and have some easier APIs for accessing its contents?

Reply:
That is a good point. I found parsing the array very error-prone myself, even while implementing it. Let me think about an easier API for digesting this array.

Reply:
In my latest changes I've added a mechanism to parse the array that I think is a bit easier to understand. This had the advantage of being easier to unit test, too.